polygeist/lib/polygeist/Ops.cpp

```diff
@@ -1912,16 +1913,14 @@ OpFoldResult Pointer2MemrefOp::fold(ArrayRef<Attribute> operands) {
     return result();
   }
   if (auto mc = source().getDefiningOp<LLVM::GEPOp>()) {
-    for (auto idx : mc.getIndices()) {
-      assert(idx);
-      if (!matchPattern(idx, m_Zero()))
+    const LLVM::GEPIndicesAdaptor<ValueRange> &indices = mc.getIndices();
+    for (auto &iter : llvm::enumerate(indices)) {
+      if (indices.isDynamicIndex(iter.index()))
+        return nullptr;
+      if (!isa<IntegerAttr>(iter.value()))
+        return nullptr;
+      if (!cast<IntegerAttr>(iter.value()).getValue().isZero())
         return nullptr;
-    }
-    auto staticIndices = mc.getStructIndices().getValues<int32_t>();
-    for (auto pair : llvm::enumerate(staticIndices)) {
-      if (pair.value() != LLVM::GEPOp::kDynamicIndex)
-        if (pair.value() != 0)
-          return nullptr;
```

Comment:
It seems to used to return nullptr when index is not zero and when struct index is dynamic and not zero.
How is the new code equivalent? The new code seems to return nullptr when the index is not an integer?

Reply:
The old code is attempting to fold a GEP when all indexes are zero. The new MLIR implementation of `getIndices()` retrieves all indices of the GEP operation, no matter if they are constant literals (e.g. attributes) or values. 

This is the new code:
```
    for (auto &iter : llvm::enumerate(indices)) {
      if (indices.isDynamicIndex(iter.index()))
        return nullptr;
      if (!isa<IntegerAttr>(iter.value()))
        return nullptr;
      if (!cast<IntegerAttr>(iter.value()).getValue().isZero())
        return nullptr;
 ```

If any index is dynamic it does not fold, if all indexes are literals but not integers do not fold, if they are all literals, integers but any is not zero also do not fold. It only folds if all indices are integer attributes with zero value.